apis/src/main/java/org/apache/rocketmq/apis/producer/Producer.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.rocketmq.apis.exception.NetworkTimeoutException;
+import org.apache.rocketmq.apis.exception.TopicDoesNotExistException;
+import org.apache.rocketmq.apis.exception.TransactionCheckerNotSetException;
+import org.apache.rocketmq.apis.message.Message;
+import org.apache.rocketmq.apis.message.MessageId;
+
+/**
+ * Producer is a thread-safe rocketmq client which is used to publish messages.
+ *
+ * <p>On account of network timeout or other reasons, rocketmq producer only promised the at-least-once semantics.
+ * For producer, at-least-once semantics means potentially attempts are made at sending it, messages may be
+ * duplicated but not lost.
+ */
+public interface Producer extends Closeable {
+    /**
+     * Sends a message synchronously.
+     *
+     * <p>This method does not return until it gets the definitive result.
+     *
+     * @param message message to send.
+     * @return the message id assigned to the appointed message.
+     * @throws TopicDoesNotExistException       if the topic of message does not exist.
+     * @throws AuthorisationException           if no permission to send message.
+     * @throws AuthenticationException          if identification could not be recognized by server.
+     * @throws ProducerClosedAlreadyException   if producer is closed already.
+     * @throws MessageTypeDoesNotMatchException if message type does not match with the topic.
+     * @throws NetworkTimeoutException          if encountered network timeout to communicate with server.
+     * @throws NetworkConnectionException       if there is a network connection problem.
+     * @throws PersistenceException             if encountered persistence failure from server.
+     */
+    MessageId send(Message message) throws ClientException;
+
+    /**
+     * Sends a transactional message synchronously.
+     *
+     * @param message     message to send.
+     * @param transaction transaction to bind.
+     * @return the message id assigned to the appointed message.
+     * @throws TopicDoesNotExistException        if the topic of message does not exist.
+     * @throws AuthorisationException            if no permission to send message.
+     * @throws AuthenticationException           if identification could not be recognized by server.
+     * @throws ProducerClosedAlreadyException    if producer is closed already.
+     * @throws MessageTypeDoesNotMatchException  if message type does not match with the topic.
+     * @throws NetworkTimeoutException           if encountered network timeout to communicate with server.
+     * @throws NetworkConnectionException        if there is a network connection problem.
+     * @throws PersistenceException              if encountered persistence failure from server.
+     * @throws TransactionCheckerNotSetException if {@link TransactionChecker} is not set.
+     */
+    MessageId send(Message message, Transaction transaction) throws ClientException;
```

Comment:
This is not the two-phase API!

How about like this :
try {
TransactionMark  mark = producer.prepareSend(Message message);
//do process
producer.commit(mark);
} catch(Exception e) {
  producer.rollback(mark);
}

Such style is easy to be integrated with db transaction. And the code will not be split.

Reply:
Well, we may send more than one message in single transaction.

Reply:
How to send more than one message using such API?
MessageId send(Message message, Transaction transaction)


The old API  for the transaction is not naturally two-phase based. It's better to polish it in the new API.
But currently, the new API is just like the old

Reply:
The code example has been added to the javadocs.

Reply:
> How to send more than one message using such API? MessageId send(Message message, Transaction transaction)
> 
> The old API for the transaction is not naturally two-phase based. It's better to polish it in the new API. But currently, the new API is just like the old

The code example has been added to the javadocs.